### ISSUE TITLE
Misleading example for joint matrix usage

### ISSUE THREAD

**Issue by javagl** [javagl/gltfOverview#10]
Originally posted at https://github.com/KhronosGroup/glTF/issues/1676 - full quote here:

> the gltf overview in the section "Combining the joint matrices to create the skinning matrix" as well as the skinning example both suggest a reference implementation using OpenGL and glsl (it is not explicitly stated though).
>
> But accessing a uniform or shader storage buffer or array using vertex attributes as indices leads to unexpected behaviour in general, because the input data for the array access is not a dynamically uniform expression (i.e. can change within the invocation) as far as i understand.
Coincidentally in the skinning example (second link) the joint indices are the same for all vertices, so in this case it could work, but that is not guaranteed.
>
> Either these examples should be somehow adapted to work in general or deleted. I think it will lead to a lot of headache for graphics programmers.

**Comment by javagl** [comment]
I'm not so deeply familiar with some points, so am not sure whether I understood the issue correctly. The point seems to be what also is described in the tutorial at https://github.com/KhronosGroup/glTF-Tutorials/blob/master/gltfTutorial/gltfTutorial_020_Skins.md#the-skinning-joints-and-weights , namely the way how the skin matrix is computed in the shader:

    attribute vec4 a_joint;
    ...
    mat4 skinMat =
        a_weight.x * u_jointMat[int(a_joint.x)] +
        a_weight.y * u_jointMat[int(a_joint.y)] +
        a_weight.z * u_jointMat[int(a_joint.z)] +
        a_weight.w * u_jointMat[int(a_joint.w)]; 

And if I understood this correctly, then this refers, roughly speaking, to some sort of possible "race condition". You mentioned that it is not a "dynamically uniform expression" - does that mean that *during* the execution of the shader, the value of `a_joint.x` could change? (That sounds strange for me, but I that's one technical point that I'm not sure about).

If so: How should the `u_jointMat` array be accessed, based on the joint indices attribute?

**Comment by infobeisel** [comment]
I think it means, for this example to work on all machines that fulfill the OpenGL spec (and nothing more), all vertices that are being rendered by the current rendering command (glDraw***) have to have **the same values in a_joint**. I say "nothing more" because for sure some implementations would still get it correct even if they are not the same, but others wouldn't.

As far as i understand, in the OpenGL context one sometimes has to consider certain data being dynamically uniform, which means, the given data does not change within the invocation group. For example, to access a uniform array, the index has to be dynamically uniform. If we have a vertex attribute called a_joint that contains the indices to access the u_jointMat array and that could possibly change for each vertex (some vertices are affected by joint 2 and 3 and some by joint 4 and 5), this condition is violated.
"Dynamically uniform expression" and "invocation group" are both rather modern (post Opengl 3x) terms as far as i can tell. "Invocation group" roughly means "the shader stage executions processed together", for compute shaders this translates to a work group, and for a vertex shader this could mean anything between "all vertices in the mesh" and "one triangle" (so a group of 3 vertices), depending on the implementation. 

To come to an end, i think it is impossible to access the `u_jointMat ` based on the `a_joint` vertex attribute. What i will do for my code (i think) is to make sure to group all vertices that have the same joint indices into meshes, and then render those. Weights on the other hand can stay vertex attributes as they are only multiplied and don't serve as array indices. Maybe this approach could be updated into the example. All this of course needs to be confirmed by an expert (i haven't been participating to any opengl spec contents or similar so far).

**EDIT:**
[this](https://www.khronos.org/opengl/wiki/Data_Type_(GLSL)#Opaque_types) and the following sections show me that the array used in the example does not need to be accessed by a dynamically uniform expression but by arbitrary values.
so i think i got it wrong, sorry. It would only be a problem if the accessed array would be an opaque array (texture samplers, atomic counters, image load/store and such)

**Comment by javagl** [comment]
Again, I did not understand all the technical details here - it has been a while since I read some details of the OpenGL spec, and many of the newer concepts (like "Dynamically uniform expression") seem to have *very* specific semantics on a very low technical level. When I wrote the overview, I read a lot about shaders and skinning in general, and think that the pattern of how the matrix is computed in the shader seemed to be how it was done everywhere, even in the glTF 1.0 sample models, where the shader was still part of the asset.

So does your **EDIT** mean that this issue is resolved? Otherwise, I'd ping some of the people who are more familiar with spec details and shaders, maybe they could help here...
